src/Storage/Device/S3.php

```diff
@@ -189,8 +285,295 @@ public function getPartitionFreeSpace():float
+
+        // Request types
+        switch ($verb) {
+            case 'GET':break;
+            case 'PUT':case 'POST': // POST only used for CloudFront
+                curl_setopt($curl, CURLOPT_CUSTOMREQUEST, $verb);
+                curl_setopt($curl, CURLOPT_POSTFIELDS, $data);
+                break;
+            case 'HEAD':
+                curl_setopt($curl, CURLOPT_CUSTOMREQUEST, 'HEAD');
+                curl_setopt($curl, CURLOPT_NOBODY, true);
+                break;
+            case 'DELETE':
+                curl_setopt($curl, CURLOPT_CUSTOMREQUEST, 'DELETE');
+                break;
+            default:break;
+        }
+
+        // Execute, grab errors
+        if (curl_exec($curl)) {
+            $this->response->code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+        } else {
+            $this->response->error = array(
+                'code' => curl_errno($curl),
+                'message' => curl_error($curl),
+                'resource' => $uri,
+            );
+        }
+
+        @curl_close($curl);
+
+        // Parse body into XML
+        if ($this->response->error === false && isset($this->response->headers['type']) &&
+            $this->response->headers['type'] == 'application/xml' && isset($this->response->body)) {
+            $this->response->body = simplexml_load_string($this->response->body);
+
+            // Grab S3 errors
+            if (!in_array($this->response->code, array(200, 204, 206)) &&
+                isset($this->response->body->Code, $this->response->body->Message)) {
+                $this->response->error = array(
+                    'code' => (string) $this->response->body->Code,
+                    'message' => (string) $this->response->body->Message,
+                );
+                if (isset($this->response->body->Resource)) {
+                    $this->response->error['resource'] = (string) $this->response->body->Resource;
+                }
+
+                unset($this->response->body);
+            }
+        }
+        return $this->response;
+    }
+
+    /**
+     * Sort compare for meta headers
+     *
+     * @internal Used to sort x-amz meta headers
+     * @param string $a String A
+     * @param string $b String B
+     * @return integer
+     */
+    private function sortMetaHeadersCmp($a, $b)
+    {
+        $lenA = strlen($a);
+        $lenB = strlen($b);
+        $minLen = min($lenA, $lenB);
+        $ncmp = strncmp($a, $b, $minLen);
+        if ($lenA == $lenB) {
+            return $ncmp;
+        }
+
+        if (0 == $ncmp) {
+            return $lenA < $lenB ? -1 : 1;
+        }
+
+        return $ncmp;
+    }
+
+    /**
+     * CURL write callback
+     *
+     * @param resource &$curl CURL resource
+     * @param string &$data Data
+     * @return integer
+     */
+    private function responseWriteCallback(&$curl, &$data)
+    {
+        $this->response->body .= $data;
+        return strlen($data);
+    }
```

Comment:
This all function can be just be passed as a callback on line 437, no real need for it to be a method.